compiler/rustc_const_eval/src/interpret/place.rs

```diff
@@ -1011,7 +1020,10 @@ where
     ) -> InterpResult<'tcx> {
         // This must be an enum or generator.
         match dest.layout.ty.kind() {
-            ty::Adt(adt, _) => assert!(adt.is_enum()),
+            ty::Adt(adt, _) => {
+                assert!(adt.is_enum());
+                self.write_uninit(dest)?;
```

Comment:
Oh, so the new semantics only applies to enums, not generators? I guess generators change their discriminant and need to preserve existing data? This should be clarified also in the `SetDiscriminant` docs.

But honestly I see this as a warning sign that maybe this is a bad idea -- if we have to treat enums and generators differently (generally, if we have to look at the *type* and not just the layout to determine MIR behavior), something is off.